jvm/src/common/Language/Java/Unsafe.hs

```diff
@@ -273,30 +277,51 @@ instance Coercible (Choice.Choice a) where
   coerce = coerce . Choice.toBool
   unsafeUncoerce = Choice.fromBool . unsafeUncoerce
 
+-- | Inject a value (of primitive or reference type) to a 'JValue'. This
+-- datatype is useful for e.g. passing arguments as a list of homogeneous type.
+-- Synonym for 'coerce'.
+jvalue :: (ty ~ Ty a, Coercible a) => a -> JValue
+jvalue = coerce
+
+-- | If @ty@ is a reference type, then it should be possible to get an object
+-- from a value.
+jobject :: (ty ~ Ty a, Coercible a, IsReferenceType ty) => a -> J ty
+jobject x
+  | JObject jobj <- coerce x = unsafeCast jobj
+  | otherwise = error "impossible"
+
 -- | Get the Java class of an object or anything 'Coercible' to one.
 classOf
   :: forall a sym. (Ty a ~ 'Class sym, Coercible a, KnownSymbol sym)
   => a
   -> JNI.String
 classOf x = JNI.fromChars (symbolVal (Proxy :: Proxy sym)) `const` coerce x
 
+-- Unexported type classes.
+class New r where
```

Comment:
You might want to consider exporting the class (but not the method). This will improve the Haddocks (where the class won't need to be written qualified) and clean up error messages that might happen to mention this class. I agree that it would be nice to avoid such error messages, but I'm sure some will slip through. The documentation should explain that it's an internal class, exported only for Haddocking and error messages. Ditto for other internal classes.

Reply:
If the class does end up leaking in error messages, then it would be better to export, I agree. The downside is that the documentation will end up bloated with `32 * 3 = 96` instances that aren't very interesting to the user. That's why I was trying to keep the class completely unexported. Interestingly, I find it pretty hard to get error messages that mention them. I haven't managed so far.

Reply:
Hm, so when the return type is wrong, type errors frequently arise elsewhere before the compiler complains about lack of instances of class `Call`. But not always. So indeed, `Call` is leaking.

Reply:
And, regardless, it will appear if a user says `:t call` in GHCi. Haskellers who see a variadic function will know that some cleverness lurks nearby, and I don't think they'll be too scared off. There might be a way to convince Haddock not to include the instances. On the other hand, seeing the instances will make it very clear what constraints will need to be satisfied by the arguments to `call`, if a user wants to take the time to piece all of this together.